client/my-sites/marketing/tools/index.tsx

```diff
@@ -15,32 +15,51 @@ import { getSelectedSiteSlug } from 'state/ui/selectors';
 import MarketingToolsFeature from './feature';
 import MarketingToolsHeader from './header';
 import { marketingSharingButtons, marketingTraffic } from 'my-sites/marketing/paths';
+import PageViewTracker from 'lib/analytics/page-view-tracker';
+import { recordTracksEvent as recordTracksEventAction } from 'state/analytics/actions';
 
 /**
  * Style dependencies
  */
 import './style.scss';
 
 interface MarketingToolsProps {
+	recordTracksEvent: ( name: string, properties: {} ) => void;
```

Comment:
I'd like to avoid this here but you could possibly convince me otherwise. at this point we're constraining the called function at the site of the callee, which feels a bit inverted.

in normal TypeScript cod this wouldn't be a problem because we'd add types to `recordTracksEvent` and that would be that. here in this hybrid we're left with partial soundness.

my recommendation in these situations is to relax the constraint and let TypeScript infer it. in one sense we're saying "this component expect this function with this type so make sure to give it" but in another sense we're really coding to `recordTracksEvent` and its API is leading our use in this function. that is, if the type signature for `recordTracksEvent` changes we want the errors to flow from that component downward. this local type constraint means that we're pushing the types upward and to me I think that's going to be less helpful in the long run.

probably the errors would be slightly different:
 - change `recordTracksEvent()`, relaxed constraint -> "these functions are calling with the wrong parameters"
 - change `recordTracksEvent()`, these constraints -> "cannot pass `recordTracksEvent` into `( string, {} ) => void`

additionally I'd hate to place the demand on everyone to manually duplicate the type definition of `recordTracksEvent` everywhere it's used. that seems like a heavy maintenance ask

Reply:
Thank you for your feedback @dmsnell. I'm going to relax that constraint, assuming by relaxing you mean switching to:

```javascript
interface Props {
  recordTracksEvent: () => void;
  selectedSiteSlug: string | null;
}
```

For the sake of learning more about TypeScript, how would adding types to the `recordTracksEvent()` function work exactly? I'm assuming it would mean:

1. Renaming [`actions.js`](https://github.com/Automattic/wp-calypso/blob/92ad10cafe535cc7892293972d5986c4504e1ff7/client/state/analytics/actions.js) in to `actions.ts`
2. Add types to `recordTracksEvent()` like this:
```javascript
export const recordTracksEvent = ( name: string, properties?: {} ) => void
  recordEvent( 'tracks', { name, properties } );
```

I'm not sure though how the `Props` interface in our `MarketingTools` component should look like in that case.  Should we keep the `recordTracksEvent: () => void;` definition?

Finally, if we switch a file to `.ts` do we have to add types to each and every function of this file? I understand this would be a better approach but obviously we all have to deal with time constraints and priorities 🙂.

Reply:
> how would adding types to the recordTracksEvent() function work exactly?

You're assuming correctly.

> if we switch a file to .ts do we have to add types to each and every function of this file?

Central to our hybrid rollout is that you don't have to add types anywhere but where they are helpful and pertinent. You can add types to `recordTracksEvent` and leave the others untouched.

> how the Props interface in our MarketingTools component should look like in that case

this is worth playing around with. if adding it explicitly to `Props` then I guess we should be doing something like `recordTracksEvent: typeof recordTracksEvent;` but I think there are ways to [reduce the overhead](https://www.typescriptlang.org/play/index.html#src=const%20decay%20%3D%20(%7B%20name%20%7D)%20%3D%3E%20%60old%20%24%7Bname%7D%60%3B%0D%0A%0D%0Atype%20Food%20%3D%20%7B%0D%0A%20%20%20%20name%3A%20string%3B%0D%0A%20%20%20%20family%3A%20'fruit'%20%7C%20'vegetable'%20%7C%20'grain'%3B%0D%0A%7D%20%26%20%7B%20decay%20%7D%0D%0A%0D%0Aconst%20f%3A%20Food%20%3D%20%7B%0D%0A%20%20%20%20name%3A%20'Apple'%2C%0D%0A%20%20%20%20family%3A%20'fruit'%2C%0D%0A%20%20%20%20decay%0D%0A%7D) of typing that out.

```ts
type Props = {
  selectedSiteSlug: string | null,
} & { recordTracksEvent }
```

This is one spot where I think splitting up `mapStateToProps` and `mapDispatchToProps` into `ConnectedProps` can be advantageous to us from a code point-of-view but YMMV